mesonbuild/interpreterbase.py

```diff
@@ -228,6 +229,124 @@ def wrapped(*wrapped_args: T.Any, **wrapped_kwargs: T.Any) -> T.Any:
+                   min_varargs: int = 0, max_varargs: int = 0) -> T.Callable[..., T.Any]:
+    """Decorator that types type checking of positional arguments.
+
+    This supports two different models of optional aguments, the first is the
+    variadic argument model. Variadic arguments are a possibly bounded,
+    possibly unbounded number of arguments of the same type (unions are
+    supported). The second is the standard default value model, in this case
+    a number of optional arguments may be provided, but they are still
+    ordered, and they may have different types.
+
+    This function does not support mixing variadic and default arguments.
+
+    :name: The name of the decorated function (as displayed in error messages)
+    :varargs: They type(s) of any variadic arguments the function takes. If
+        None the function takes no variadic args
+    :min_varargs: the minimum number of variadic arguments taken
+    :max_varargs: the maximum number of variadic arguments taken. 0 means unlimited
+    :optargs: The types of any optional arguments parameters taken. If None
+        then no optional paramters are taken.
+
+    allows replacing this:
+    ```python
+    def func(self, node, args, kwargs):
+        if len(args) != 2:
+            raise Exception('... takes exactly 2 arguments)
+        foo: str = args[0]
+        if not isinstance(foo, str):
+            raise ...
+        bar: int = args[1]
+        if not isinstance(bar, int):
+            raise ...
+
+        # actual useful stuff
+    ```
+    with:
+    ```python
+    @typed_pos_args('func_name', str, int)
+    def func(self, node, args: T.Tuple[str, int], kwargs):
+        foo, bar = args
+
+        # actual useful stuff
+    ```
+    """
+    def inner(f: TV_func) -> TV_func:
+
+        @wraps(f)
+        def wrapper(*wrapped_args: T.Any, **wrapped_kwargs: T.Any) -> T.Any:
+            args = _get_callee_args(wrapped_args)[2]
+
+            # These are implementation programming errors, end users should never see them.
+            assert isinstance(args, list), args
+            assert max_varargs >= 0, 'max_varrags cannot be negative'
+            assert min_varargs >= 0, 'min_varrags cannot be negative'
+            assert optargs is None or varargs is None, \
+                'varargs and optargs not supported together as this would be ambiguous'
+
+            num_args = len(args)
+            num_types = len(types)
+            a_types = types
+
+            if varargs:
+                num_types += min_varargs
+                if max_varargs == 0 and num_args < num_types:
+                    raise InvalidArguments(f'{name} takes at least {num_types} arguments, but got {num_args}.')
+                elif max_varargs != 0 and (num_args < num_types or num_args > num_types + max_varargs - min_varargs):
+                    raise InvalidArguments(f'{name} takes between {num_types} and {num_types + max_varargs - min_varargs} arguments, but got {len(args)}.')
```

Comment:
I find it easier to read by adding more descriptive variables:
```suggestion
                min_args = num_types + min_varargs
                max_args = num_types + max_varargs
                if max_varargs == 0 and num_args < min_args:
                    raise InvalidArguments(f'{name} takes at least {min_args} arguments, but got {num_args}.')
                elif max_varargs != 0 and (num_args < min_args or num_args > max_args)):
                    raise InvalidArguments(f'{name} takes between {min_args} and {max_args} arguments, but got {num_args}.')
```

What do you think?

Reply:
Seems good to me. I've updated that and updated the docstring to try to be more useful.